File: src/portroyale/sim.py
"""Monte Carlo simulation engine.

A *session* is one run of a strategy: start with a bankroll, let the
strategy bet before every round, resolve the round, and repeat for ``rolls``
rounds (or until the bankroll can no longer cover the table minimum —
that's a ruin). ``run_simulation`` repeats the session ``sessions`` times,
optionally across multiple processes, and aggregates everything into a
:class:`SimReport` with the statistics that matter for a strategy:

- realized house edge (net loss per dollar wagered — the number the math
  says your strategy *actually* pays),
- session win rate, risk of ruin, max drawdown,
- percentiles of final bankrolls, plus percentile bands over time for
  charts.
"""
from __future__ import annotations

import os
from concurrent.futures import ProcessPoolExecutor
from dataclasses import dataclass, field
from typing import Any, Optional

import numpy as np

from .games import make_table
from .games.base import InsufficientBankroll, TableError
from .strategies import make_strategy

HISTORY_POINTS = 400  # bankroll samples kept per session (for charts)


@dataclass
class TableConfig:
    min_bet: float = 5.0
    max_bet: float = 5000.0
    odds_multiple: int = 5


@dataclass
class SimConfig:
    game: str = "craps"
    strategy: str = "passline_odds"
    strategy_params: dict = field(default_factory=dict)
    bankroll: float = 1000.0
    table: TableConfig = field(default_factory=TableConfig)
    rolls: int = 10_000
    sessions: int = 500
    seed: Optional[int] = None
    workers: int = 0  # 0 -> os.cpu_count()

    def as_dict(self) -> dict:
        return {
            "game": self.game,
            "strategy": self.strategy,
            "strategy_params": dict(self.strategy_params),
            "bankroll": self.bankroll,
            "table": {
                "min_bet": self.table.min_bet,
                "max_bet": self.table.max_bet,
                "odds_multiple": self.table.odds_multiple,
            },
            "rolls": self.rolls,
            "sessions": self.sessions,
            "seed": self.seed,
        }


@dataclass
class SessionResult:
    final_bankroll: float
    profit: float
    action: float          # total dollars wagered
    edge: float            # realized house edge for this session (-profit/action)
    max_drawdown: float
    ruined: bool
    rolls_played: int
    history: list[float]   # downsampled bankroll trajectory


def run_session(cfg: dict, session_index: int) -> SessionResult:
    """Run one session. Module-level so it pickles for multiprocessing."""
    seed = cfg["seed"]
    rng = np.random.default_rng() if seed is None else np.random.default_rng(seed + session_index)
    t = cfg["table"]
    table = make_table(
        cfg["game"],
        bankroll=cfg["bankroll"],
        min_bet=t["min_bet"],
        max_bet=t["max_bet"],
        odds_multiple=t["odds_multiple"],
        rng=rng,
    )
    strategy = make_strategy(cfg["strategy"], **cfg["strategy_params"])
    strategy.on_session_start(table)

    start = cfg["bankroll"]
    # Sample at most HISTORY_POINTS points; pad short runs to equal length.
    step = max(1, -(-cfg["rolls"] // HISTORY_POINTS))  # ceiling division
    history = [start]
    peak = start
    max_dd = 0.0
    ruined = False
    rolls_played = 0

    for r in range(1, cfg["rolls"] + 1):
        if table.bankroll < table.min_bet - 1e-9:
            ruined = True
            break
        try:
            strategy.decide(table)
            table.roll()
        except (InsufficientBankroll, TableError):
            # Strategy bug or broke mid-round: stop the session here.
            ruined = table.bankroll < table.min_bet
            break
        rolls_played = r
        peak = max(peak, table.bankroll)
        max_dd = max(max_dd, peak - table.bankroll)
        if r % step == 0 or r == cfg["rolls"]:
            history.append(table.bankroll)

    # Pad the trajectory so every session lines up for percentile bands.
    while len(history) < HISTORY_POINTS + 1:
        history.append(history[-1])

    profit = table.bankroll - start
    action = table.total_wagered
    edge = -profit / action if action > 0 else 0.0
    return SessionResult(
        final_bankroll=table.bankroll,
        profit=profit,
        action=action,
        edge=edge,
        max_drawdown=max_dd,
        ruined=ruined,
        rolls_played=rolls_played,
        history=history,
    )


@dataclass
class SimReport:
    config: SimConfig
    sessions: int
    mean_final: float
    median_final: float
    p5_final: float
    p25_final: float
    p75_final: float
    p95_final: float
    mean_profit: float
    total_action: float
    realized_edge: float      # -total_profit / total_action
    win_rate: float           # fraction of sessions finishing above start
    ruin_rate: float          # fraction of sessions ruined
    mean_max_drawdown: float
    median_rolls_played: float
    bands: dict[str, list[float]]  # p5/p25/p50/p75/p95 bankroll over time


def run_simulation(config: SimConfig, progress: Any = None) -> SimReport:
    """Run ``config.sessions`` sessions, in parallel, and aggregate."""
    cfg = config.as_dict()
    workers = config.workers or os.cpu_count() or 1

    if workers == 1:
        results = [run_session(cfg, i) for i in range(config.sessions)]
    else:
        with ProcessPoolExecutor(max_workers=workers) as pool:
            results = list(pool.map(run_session, [cfg] * config.sessions,
                                    range(config.sessions)))

    finals = np.array([r.final_bankroll for r in results])
    profits = np.array([r.profit for r in results])
    actions = np.array([r.action for r in results])
    histories = np.array([r.history for r in results])  # (sessions, points)

    total_profit = float(profits.sum())
    total_action = float(actions.sum())
    bands = {
        "p5": np.percentile(histories, 5, axis=0).tolist(),
        "p25": np.percentile(histories, 25, axis=0).tolist(),
        "p50": np.percentile(histories, 50, axis=0).tolist(),
        "p75": np.percentile(histories, 75, axis=0).tolist(),
        "p95": np.percentile(histories, 95, axis=0).tolist(),
    }
    return SimReport(
        config=config,
        sessions=len(results),
        mean_final=float(finals.mean()),
        median_final=float(np.median(finals)),
        p5_final=float(np.percentile(finals, 5)),
        p25_final=float(np.percentile(finals, 25)),
        p75_final=float(np.percentile(finals, 75)),
        p95_final=float(np.percentile(finals, 95)),
        mean_profit=float(profits.mean()),
        total_action=total_action,
        realized_edge=(-total_profit / total_action) if total_action > 0 else 0.0,
        win_rate=float(np.mean(finals > config.bankroll)),
        ruin_rate=float(np.mean([r.ruined for r in results])),
        mean_max_drawdown=float(np.mean([r.max_drawdown for r in results])),
        median_rolls_played=float(np.median([r.rolls_played for r in results])),
        bands=bands,
    )


def format_report(report: SimReport) -> str:
    """Render a SimReport as a plain-text summary (used by the CLI)."""
    c = report.config
    params = ", ".join(f"{k}={v}" for k, v in c.strategy_params.items())
    lines = [
        f"Port Royale — {c.game} / {c.strategy}" + (f" ({params})" if params else ""),
        f"Sessions: {report.sessions}   Rolls/session: {c.rolls}   "
        f"Bankroll: ${c.bankroll:,.2f}",
        "-" * 60,
        f"Median final bankroll:  ${report.median_final:>12,.2f}",
        f"Mean final bankroll:    ${report.mean_final:>12,.2f}",
        f"Mean net profit:        ${report.mean_profit:>+12,.2f}",
        f"Total action:           ${report.total_action:>12,.2f}",
        f"Realized house edge:    {report.realized_edge:>11.2%}  (per $ wagered)",
        f"Session win rate:       {report.win_rate:>11.1%}",
        f"Risk of ruin:           {report.ruin_rate:>11.1%}",
        f"Mean max drawdown:      ${report.mean_max_drawdown:>12,.2f}",
        "Final bankroll percentiles:",
        f"   p5 ${report.p5_final:,.2f}   p25 ${report.p25_final:,.2f}   "
        f"p75 ${report.p75_final:,.2f}   p95 ${report.p95_final:,.2f}",
    ]
    return "\n".join(lines)

File: mobile/engine/sim.ts
/**
 * Monte Carlo simulation engine.
 *
 * Concept-for-concept port of `sim.py`. A *session* is one run of a
 * strategy: start with a bankroll, let the strategy bet before every
 * round, resolve the round, and repeat for `rolls` rounds (or until the
 * bankroll can no longer cover the table minimum — that's a ruin).
 *
 * Reproducibility: every session is fully determined by its seed. The
 * master seed plus the session index derives the session seed
 * (`master + index`), exactly like the Python engine. Note the dice
 * *values* differ from Python for the same seed (different PRNG — see
 * `prng.ts`); determinism is per-platform.
 *
 * Watch mode: `watchSession` plays a single session and yields the
 * JSON-serializable event stream defined in `events.ts` (the
 * cross-platform contract — see `docs/watch-mode.md`). `runSession` is
 * built on the same generator, so simulated statistics and the visible
 * play-by-play can never disagree.
 *
 * Comparison mode: `runComparison` runs several strategies against the
 * *identical* dice streams (common random numbers).
 *
 * UI-free: no React / React Native imports.
 */

import { CrapsTable } from "./craps";
import {
  BankrollEvent,
  BetPlacedEvent,
  BetResolvedEvent,
  RollEvent,
  SessionEndEvent,
  SessionEndReason,
  SessionStartEvent,
  WatchEvent,
} from "./events";
import { SeededRng } from "./prng";
import { makeStrategy } from "./strategies";
import { Bet, BetEvent, TableError } from "./table";

export const HISTORY_POINTS = 400; // bankroll samples kept per session

export const REASON_EXHAUSTED = "rolls_exhausted" as const;
export const REASON_RUIN = "ruin" as const;
export const REASON_STOP_LOSS = "stop_loss" as const;
export const REASON_STOP_WIN = "stop_win" as const;
export const REASON_ERROR = "error" as const;

export interface TableConfig {
  minBet: number;
  maxBet: number;
  oddsMultiple: number;
}

export const DEFAULT_TABLE_CONFIG: TableConfig = {
  minBet: 5.0,
  maxBet: 5000.0,
  oddsMultiple: 5,
};

/** Derive a session's seed deterministically from the master seed. */
export function sessionSeed(
  masterSeed: number | null | undefined,
  sessionIndex: number
): number | null {
  return masterSeed == null ? null : masterSeed + sessionIndex;
}

/** Generate `n` dice pairs from `seed`. Shared by sim and compare. */
export function makeDiceStream(seed: number, n: number): Array<[number, number]> {
  const rng = new SeededRng(seed);
  const out: Array<[number, number]> = [];
  for (let i = 0; i < n; i++) out.push([rng.die(), rng.die()]);
  return out;
}

/* ------------------------------------------------------------------ */
/* Session event generator (the single source of truth)                */
/* ------------------------------------------------------------------ */

export interface WatchOptions {
  game?: string;
  strategy?: string;
  strategyParams?: Record<string, unknown>;
  bankroll?: number;
  minBet?: number;
  maxBet?: number;
  oddsMultiple?: number;
  rolls?: number;
  seed?: number | null;
  stopLoss?: number | null; // end session at <= this fraction of start
  stopWin?: number | null; // end session at >= this fraction of start
  dice?: Array<[number, number]> | null;
}

function betPlacedEvent(rollNo: number, bet: Bet): BetPlacedEvent {
  return {
    type: "bet_placed",
    roll: rollNo,
    kind: bet.kind,
    amount: bet.amount,
    number: bet.number,
  };
}

function betResolvedEvent(rollNo: number, event: BetEvent): BetResolvedEvent {
  return {
    type: "bet_resolved",
    roll: rollNo,
    kind: event.kind,
    number: event.number,
    amount: event.amount,
    outcome: event.outcome,
    profit: event.profit,
  };
}

function makeTable(
  game: string,
  opts: {
    bankroll: number;
    minBet: number;
    maxBet: number;
    oddsMultiple: number;
  }
): CrapsTable {
  if (game !== "craps") {
    throw new Error(`unknown game ${JSON.stringify(game)}; available: ["craps"]`);
  }
  return new CrapsTable(opts);
}

/**
 * Play one session, yielding JSON-serializable event dicts.
 *
 * Per-roll ordering (the contract): bet_placed × N, bet_resolved × M
 * (take-downs during decide), roll, bet_resolved × K (settlements),
 * bankroll.
 */
export function* watchSession(
  opts: WatchOptions = {},
  sessionIndex = 0,
  dice: Array<[number, number]> | null = null
): Generator<WatchEvent> {
  const game = opts.game ?? "craps";
  const strategyName = opts.strategy ?? "passline_odds";
  const strategyParams = { ...(opts.strategyParams ?? {}) };
  const bankroll = opts.bankroll ?? 1000.0;
  const minBet = opts.minBet ?? DEFAULT_TABLE_CONFIG.minBet;
  const maxBet = opts.maxBet ?? DEFAULT_TABLE_CONFIG.maxBet;
  const oddsMultiple = opts.oddsMultiple ?? DEFAULT_TABLE_CONFIG.oddsMultiple;
  const rolls = opts.rolls ?? 300;
  const masterSeed = opts.seed ?? 42;
  const stopLoss = opts.stopLoss ?? null;
  const stopWin = opts.stopWin ?? null;

  if (stopLoss != null && stopLoss <= 0) {
    throw new Error(`stopLoss must be positive, got ${stopLoss}`);
  }
  if (stopWin != null && stopWin <= 0) {
    throw new Error(`stopWin must be positive, got ${stopWin}`);
  }

  const seed = sessionSeed(masterSeed, sessionIndex);
  const explicitDice = opts.dice ?? dice;
  const stream =
    explicitDice ?? (seed != null ? makeDiceStream(seed, rolls) : null);

  const table = makeTable(game, { bankroll, minBet, maxBet, oddsMultiple });
  const strategy = makeStrategy(strategyName, strategyParams);
  strategy.onSessionStart(table);

  const start = bankroll;

  const startEvent: SessionStartEvent = {
    type: "session_start",
    game,
    strategy: strategyName,
    strategy_params: strategyParams,
    seed,
    bankroll: start,
    rolls,
    stop_loss: stopLoss,
    stop_win: stopWin,
    min_bet: minBet,
    max_bet: maxBet,
    odds_multiple: Math.trunc(oddsMultiple),
  };
  yield startEvent;

  let peak = start;
  let maxDD = 0;
  let rollsPlayed = 0;
  let reason: SessionEndReason = REASON_EXHAUSTED;
  let error: string | undefined;

  for (let r = 1; r <= rolls; r++) {
    if (table.bankroll < table.minBet - 1e-9) {
      reason = REASON_RUIN;
      break;
    }
    const known = new Set<Bet>(table.bets);
    try {
      strategy.decide(table);
    } catch (e) {
      if (e instanceof TableError) {
        reason = REASON_ERROR;
        error = e.message;
        break;
      }
      throw e;
    }
    for (const bet of table.bets) {
      if (!known.has(bet)) yield betPlacedEvent(r, bet);
    }
    for (const event of table.drainEvents()) {
      // take-downs / adjustments during decide()
      yield betResolvedEvent(r, event);
    }

    let d1: number;
    let d2: number;
    if (stream) {
      [d1, d2] = stream[r - 1];
      table.roll(d1, d2);
    } else {
      const result = table.roll();
      d1 = result.d1;
      d2 = result.d2;
    }
    const rollEvent: RollEvent = {
      type: "roll",
      roll: r,
      d1,
      d2,
      total: d1 + d2,
      phase: table.phase, // phase *after* the roll resolved
      point: table.point, // point *after* the roll resolved
    };
    yield rollEvent;
    for (const event of table.drainEvents()) {
      yield betResolvedEvent(r, event);
    }

    rollsPlayed = r;
    peak = Math.max(peak, table.bankroll);
    maxDD = Math.max(maxDD, peak - table.bankroll);
    const bankrollEvent: BankrollEvent = {
      type: "bankroll",
      roll: r,
      bankroll: table.bankroll,
    };
    yield bankrollEvent;

    if (stopWin != null && table.bankroll >= stopWin * start - 1e-9) {
      reason = REASON_STOP_WIN;
      break;
    }
    if (stopLoss != null && table.bankroll <= stopLoss * start + 1e-9) {
      reason = REASON_STOP_LOSS;
      break;
    }
  }

  const endEvent: SessionEndEvent = {
    type: "session_end",
    reason,
    rolls_played: rollsPlayed,
    final_bankroll: table.bankroll,
    net_profit: table.bankroll - start,
    total_wagered: table.totalWagered,
    max_drawdown: maxDD,
  };
  if (error !== undefined) endEvent.error = error;
  yield endEvent;
}

/* ------------------------------------------------------------------ */
/* Session results                                                     */
/* ------------------------------------------------------------------ */

export interface SessionResult {
  finalBankroll: number;
  profit: number;
  action: number; // total dollars wagered
  edge: number; // realized house edge for this session (-profit/action)
  maxDrawdown: number;
  ruined: boolean;
  rollsPlayed: number;
  endReason: SessionEndReason;
  seed: number | null;
  history: number[]; // downsampled bankroll trajectory
}

/**
 * Run one session, consuming the same event generator as `watchSession`
 * so simulated statistics and the visible play-by-play always agree.
 */
export function runSession(
  cfg: Required<SimConfig>,
  sessionIndex: number,
  dice: Array<[number, number]> | null = null
): SessionResult {
  const seed = sessionSeed(cfg.seed, sessionIndex);
  const start = cfg.bankroll;
  // Sample at most HISTORY_POINTS points; pad short runs to equal length.
  const step = Math.max(1, Math.ceil(cfg.rolls / HISTORY_POINTS));
  const history: number[] = [start];
  let endReason: SessionEndReason = REASON_EXHAUSTED;
  let final = start;
  let profit = 0;
  let action = 0;
  let maxDD = 0;
  let rollsPlayed = 0;

  for (const event of watchSession(
    {
      game: cfg.game,
      strategy: cfg.strategy,
      strategyParams: cfg.strategyParams,
      bankroll: cfg.bankroll,
      minBet: cfg.table.minBet,
      maxBet: cfg.table.maxBet,
      oddsMultiple: cfg.table.oddsMultiple,
      rolls: cfg.rolls,
      seed: cfg.seed,
      stopLoss: cfg.stopLoss,
      stopWin: cfg.stopWin,
    },
    sessionIndex,
    dice
  )) {
    if (event.type === "bankroll") {
      const r = event.roll;
      if (r % step === 0 || r === cfg.rolls) history.push(event.bankroll);
    } else if (event.type === "session_end") {
      endReason = event.reason;
      final = event.final_bankroll;
      profit = event.net_profit;
      action = event.total_wagered;
      maxDD = event.max_drawdown;
      rollsPlayed = event.rolls_played;
      if (rollsPlayed % step !== 0) history.push(final);
    }
  }

  // Pad the trajectory so every session lines up for percentile bands.
  while (history.length < HISTORY_POINTS + 1) {
    history.push(history[history.length - 1]);
  }

  const edge = action > 0 ? -profit / action : 0;
  return {
    finalBankroll: final,
    profit,
    action,
    edge,
    maxDrawdown: maxDD,
    ruined: endReason === REASON_RUIN,
    rollsPlayed,
    endReason,
    seed,
    history,
  };
}

/* ------------------------------------------------------------------ */
/* Statistics                                                          */
/* ------------------------------------------------------------------ */

/** Percentile with linear interpolation — matches numpy.percentile default. */
export function percentile(sorted: number[], q: number): number {
  if (sorted.length === 0) return NaN;
  if (sorted.length === 1) return sorted[0];
  const pos = (q / 100) * (sorted.length - 1);
  const lo = Math.floor(pos);
  const hi = Math.ceil(pos);
  return sorted[lo] + (sorted[hi] - sorted[lo]) * (pos - lo);
}

function quantile(values: number[], q: number): number {
  return percentile([...values].sort((a, b) => a - b), q);
}

function mean(values: number[]): number {
  return values.reduce((s, v) => s + v, 0) / Math.max(1, values.length);
}

export interface SimReport {
  sessions: number;
  meanFinal: number;
  medianFinal: number;
  p5Final: number;
  p25Final: number;
  p75Final: number;
  p95Final: number;
  meanProfit: number;
  totalAction: number;
  /** -totalProfit / totalAction */
  realizedEdge: number;
  /** fraction of sessions finishing above start */
  winRate: number;
  /** fraction of sessions ruined */
  ruinRate: number;
  meanMaxDrawdown: number;
  medianRollsPlayed: number;
  /** reason -> fraction of sessions */
  endReasons: Record<string, number>;
  /** p5/p25/p50/p75/p95 bankroll over time */
  bands: Record<string, number[]>;
}

export interface SimConfig {
  game?: string;
  strategy?: string;
  strategyParams?: Record<string, unknown>;
  bankroll?: number;
  table?: Partial<TableConfig>;
  rolls?: number;
  sessions?: number;
  seed?: number | null;
  stopLoss?: number | null;
  stopWin?: number | null;
}

function normalizeConfig(cfg: SimConfig): Required<SimConfig> {
  const table: TableConfig = {
    ...DEFAULT_TABLE_CONFIG,
    ...(cfg.table ?? {}),
  };
  return {
    game: cfg.game ?? "craps",
    strategy: cfg.strategy ?? "passline_odds",
    strategyParams: cfg.strategyParams ?? {},
    bankroll: cfg.bankroll ?? 1000.0,
    table,
    rolls: cfg.rolls ?? 10000,
    sessions: cfg.sessions ?? 500,
    seed: cfg.seed ?? null,
    stopLoss: cfg.stopLoss ?? null,
    stopWin: cfg.stopWin ?? null,
  };
}

function aggregate(
  results: SessionResult[],
  config: Required<SimConfig>
): SimReport {
  const finals = results.map((r) => r.finalBankroll);
  const profits = results.map((r) => r.profit);
  const actions = results.map((r) => r.action);

  const totalProfit = profits.reduce((s, v) => s + v, 0);
  const totalAction = actions.reduce((s, v) => s + v, 0);

  const bands: Record<string, number[]> = {};
  for (const [name, q] of [
    ["p5", 5],
    ["p25", 25],
    ["p50", 50],
    ["p75", 75],
    ["p95", 95],
  ] as Array<[string, number]>) {
    const series: number[] = [];
    const points = results[0]?.history.length ?? 0;
    for (let i = 0; i < points; i++) {
      series.push(quantile(results.map((r) => r.history[i]), q));
    }
    bands[name] = series;
  }

  const endReasons: Record<string, number> = {};
  for (const r of results) {
    endReasons[r.endReason] = (endReasons[r.endReason] ?? 0) + 1;
  }
  const n = Math.max(1, results.length);
  for (const k of Object.keys(endReasons)) endReasons[k] /= n;

  return {
    sessions: results.length,
    meanFinal: mean(finals),
    medianFinal: quantile(finals, 50),
    p5Final: quantile(finals, 5),
    p25Final: quantile(finals, 25),
    p75Final: quantile(finals, 75),
    p95Final: quantile(finals, 95),
    meanProfit: mean(profits),
    totalAction,
    realizedEdge: totalAction > 0 ? -totalProfit / totalAction : 0,
    winRate: mean(finals.map((f) => (f > config.bankroll ? 1 : 0))),
    ruinRate: mean(results.map((r) => (r.ruined ? 1 : 0))),
    meanMaxDrawdown: mean(results.map((r) => r.maxDrawdown)),
    medianRollsPlayed: quantile(
      results.map((r) => r.rollsPlayed),
      50
    ),
    endReasons,
    bands,
  };
}

/** Run `sessions` sessions (synchronously) and aggregate into a report. */
export function runSimulation(config: SimConfig): SimReport {
  const cfg = normalizeConfig(config);
  const results: SessionResult[] = [];
  for (let i = 0; i < cfg.sessions; i++) {
    results.push(runSession(cfg, i));
  }
  return aggregate(results, cfg);
}

/* ------------------------------------------------------------------ */
/* Comparison mode (common random numbers)                             */
/* ------------------------------------------------------------------ */

export interface ComparisonReport {
  game: string;
  strategies: string[];
  sessions: number;
  rolls: number;
  bankroll: number;
  seed: number;
  reports: Record<string, SimReport>;
}

/**
 * Run several strategies against IDENTICAL dice streams.
 *
 * For session `i` the dice are generated once from `seed + i` and replayed
 * for every strategy — common random numbers, so result differences
 * reflect strategy differences, not luck. A seed is required.
 */
export function runComparison(opts: {
  game?: string;
  strategies: string[];
  strategyParams?: Record<string, Record<string, unknown>>;
  bankroll?: number;
  table?: Partial<TableConfig>;
  rolls?: number;
  sessions?: number;
  seed: number | null;
  stopLoss?: number | null;
  stopWin?: number | null;
}): ComparisonReport {
  const {
    game = "craps",
    strategies,
    strategyParams = {},
    bankroll = 1000.0,
    table,
    rolls = 5000,
    sessions = 100,
    seed,
    stopLoss = null,
    stopWin = null,
  } = opts;

  if (seed == null) {
    throw new Error(
      "compare requires a seed — identical dice are the whole point"
    );
  }
  if (strategies.length < 2) {
    throw new Error("compare needs at least two strategies");
  }

  const reports: Record<string, SimReport> = {};
  for (const name of strategies) {
    const cfg = normalizeConfig({
      game,
      strategy: name,
      strategyParams: strategyParams[name] ?? {},
      bankroll,
      table,
      rolls,
      sessions,
      seed,
      stopLoss,
      stopWin,
    });
    const results: SessionResult[] = [];
    for (let i = 0; i < sessions; i++) {
      const dice = makeDiceStream(seed + i, rolls);
      results.push(runSession(cfg, i, dice));
    }
    reports[name] = aggregate(results, cfg);
  }

  return { game, strategies, sessions, rolls, bankroll, seed, reports };
}
